Simplify single source (#1580)
pypa/packaging.python.org: source/discussions/single-source-version.rst

> .. _`Single sourcing the version`:

Reviewer: This label duplicates the one already declared in another document, which legitimately breaks the CI @ https://github.com/pypa/packaging.python.org/actions/runs/10514372639/job/29137925991?pr=1580#step:5:126 with

```console
/home/runner/work/packaging.python.org/packaging.python.org/source/guides/single-sourcing-package-version.rst:5: WARNING: duplicate label single sourcing the version, other instance in /home/runner/work/packaging.python.org/packaging.python.org/source/discussions/single-source-version.rst
```

So it has to be different. I think, we can go for
```suggestion
.. _`Single sourcing the version discussion`:
```

===================================
Single-sourcing the Project Version
===================================

:Page Status: Complete

> :Last Reviewed: 2024-??

Reviewer: TODO: enter the actual date pre-merge. *Do not mark this discussion as resolved until right before clicking merge.*

Reviewer: ```suggestion
:Last Reviewed: 2024-08-24
```

One of the challenges in building packages is that the version string can be required in multiple places.

* It needs to be specified when building the package (e.g. in :file:`pyproject.toml`)
   This will make it available in the installed package’s metadata, from where it will be accessible at runtime using ``importlib.metadata.version("distribution_name")``.

> * A package may set a top level ``__version__`` attribute to provide an alternative means of runtime access to the version of the imported package. If this is done, the value of ``__version__`` attribute and that used by the build system to set the distribution's version should be kept in sync in :ref:`the build systems's recommended way <Build system version handling>`.

Reviewer 2: Maybe say *(usually called `__version__`)* and leave it at that? Some packages also set a `__version_info__` or have separate long and short versions or ... there are lots of them. Let's just endorse the one that's more or less (albeit unofficially) seen as *standard*.

> * A package may set a top level ``__version__`` attribute to provide an alternative means of runtime access to the version of the imported package. If this is done, the value of ``__version__`` attribute and that used by the build system to set the distribution's version should be kept in sync in :ref:`the build systems's recommended way <Build system version handling>`.

Reviewer 3: I don't see why endorsing it, because I think this is a matter of API design.
I also don't think it is seen as a *standard*, more like a popular *convention* (`VERSION` is also a very well stablished convention, just not as popular)...

But I am completely OK with @bwoodsend suggestion. Something would also work for me is:

```suggestion
* A package may set a module attribute (e.g., ``__version__``) to provide an alternative means of runtime access to the version of the imported package. If this is done, the value of the attribute and that used by the build system to set the distribution's version should be kept in sync in :ref:`the build systems's recommended way <Build system version handling>`.
```

> * A package may set a top level ``__version__`` attribute to provide an alternative means of runtime access to the version of the imported package. If this is done, the value of ``__version__`` attribute and that used by the build system to set the distribution's version should be kept in sync in :ref:`the build systems's recommended way <Build system version handling>`.

Author: > [people do use VERSION](https://github.com/search?q=%2Fattr+%3D+%22.*VERSION%22%2F+language%3ATOML&type=code&l=TOML) and [other things](https://github.com/search?q=%2Fattr+%3D+%22.*%5B%5E_%22%5D%22%2F+language%3ATOML+path%3Apyproject.toml&type=code)... Probably this is best left as a design choice for the API of each individual package (if they want to expose it through their API).

Sure they do -- but while there is no consensus about whether an import module should have an attribute that returns the version string, I hope there is consensus that it's better for the community if everyone uses the same name for it, and `__version__` is the most common convention at this point. In fact, a `__version__` PEP is being revived to hopefully codify it -- but in any case, in all the discussions I've seen on this topic, `__version__` is the ONLY name I've seen advocated for (as a standard). 

In any case, I've accepted @bwoodsend 's text to soften it a bit.

> * A package may set a top level ``__version__`` attribute to provide an alternative means of runtime access to the version of the imported package. If this is done, the value of ``__version__`` attribute and that used by the build system to set the distribution's version should be kept in sync in :ref:`the build systems's recommended way <Build system version handling>`.

Reviewer: @ChrisBarker-NOAA at first, I was misunderstanding why the opposition existed — I thought it was because of the name. But having read the explanations later in the process, I realized that it was about having import-time logic.

When I brought this up in a code review of another project that relies on `asyncio`, I heard an interesting point — having heavy version computation in `__init__.py` during import-time (as opposed to having `__version__` under `__getattr__()` for lazy access) might be preferable because of the predictability, when it's known to be always used. And having it computed lazily would be blocking to the event loop at a non-deterministic point in time.

Just wanted to mention this...

That said, when I pointed out that setuptools has an option to source the version from importables through dynamic attributes, that resolved the problem even better as with that, there's no sudden computational cost to it. And that version in a Python module could be the original source of truth as the version data access flow is unidirectional — `[importable] -> [installed metadata] -> [reading from metadata]` or `[importable] -> [project-internal code reading from importable]` (which fits well in my diagram: https://github.com/pypa/packaging.python.org/pull/1276#issuecomment-1709024176). This might be useful for programs that access and display their own versions without concerning themselves of how they're injected into venv.

So I think that it might be useful to mention the cost of having `__version__` with a static value vs. `version = importlib.metadata.version(__package__)`. These cases would just have different sources of where the version comes from initially. And the external programs would be still recommended to do `importlib.metadata.version(__package__)` in general case.

Perhaps, we shouldn't present the idea of having `__version__` as an “approved/recommended public API:tm:” for external programs but reframe that as “*sometimes* this attribute exists, *sometimes* it's called `__version__`, *some* projects may use it internally and for packaging metadata”

> * A package may set a top level ``__version__`` attribute to provide an alternative means of runtime access to the version of the imported package. If this is done, the value of ``__version__`` attribute and that used by the build system to set the distribution's version should be kept in sync in :ref:`the build systems's recommended way <Build system version handling>`.

Author: Surprisingly, this is a VERY contentious topic -- I am trying desperately to capture the key point, which is:

have one source of truth for yur version string when building a package.

While avoiding recommendations on what should actually be in the final, installed package.

I think that discussion is good, but should be elsewhere.

NOTE: the current structure of these docs does, indeed, have a "discussions" section -- -so it could be that a nice full discussion of the issue would be accepted -- but let's do that as another PR :-)

> * A package may set a top level ``__version__`` attribute to provide an alternative means of runtime access to the version of the imported package. If this is done, the value of ``__version__`` attribute and that used by the build system to set the distribution's version should be kept in sync in :ref:`the build systems's recommended way <Build system version handling>`.

Reviewer: Alright, noted. I think I might've subconsciously extended the scope of your effort in my head :)

Marking this thread as resolved for now.

* If the code is in in a version control system (VCS), e.g. Git, the version may appear in a *tag* such as ``v1.2.3``.

To ensure that version numbers do not get out of sync, it is recommended that there is a single source of truth for the version number.

In general, the options are:

1) If the code is in a version control system (VCS), e.g. Git, then the version can be extracted from the VCS.

2) The version can be hard-coded into the :file:`pyproject.toml` file -- and the build system can copy it into other locations it may be required.

> 3) The version string can be hard-coded into the source code -- either in a special purpose file, such as :file:`_version.txt`, or as a attribute in the :file:`__init__.py`, and the build system can extract it at build time.

Reviewer 3: [Some people do use other modules](https://github.com/search?q=%2Fattr+%3D+%22.*VERSION%22%2F+language%3ATOML&type=code&l=TOML) to store the version (e.g., `version.py`, or in the case of single-module distributions, whichever module name happens to be the main module).

```suggestion
3) The version string can be hard-coded into the source code -- either in a special purpose file, such as :file:`_version.txt`, or as a attribute in a module, such as :file:`__init__.py`, and the build system can extract it at build time.
```

Author: You're quite right, a single file module doesn't have an `__init__.py` -- thanks!

Consult your build system's documentation for their recommended method.

.. _Build system version handling:

Build System Version Handling
-----------------------------

The following are links to some build system's documentation for handling version strings.

* `Flit <https://flit.pypa.io/en/stable/>`_

* `Hatchling <https://hatch.pypa.io/1.9/version/>`_

* `PDM <https://pdm-project.org/en/latest/reference/pep621/#__tabbed_1_2>`_

> * `Setuptools <https://setuptools.pypa.io/en/latest/userguide/distribution.html#specifying-your-project-s-version>`_

Reviewer 3: Probably the link that comes closer to the intent of the text would be https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata

```suggestion
* `Setuptools <https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata>`_
```

Author: well, yes, but that's out of scope for this particular doc ..... I may submit a PR there, if I can find the time -- if anyone else gets to it first -- great! 

Hmm -- this doc is not necessarily endorsing the dynamic method per se -- but sure, that's probably a better link. Accepted.

Reviewer: @ChrisBarker-NOAA is it possible to use intersphinx instead of a hardcoded link? I'm pretty sure that setuptools' docs are plugged into Sphinx already.

Author: probably -- if you can figure out how, I'll accept the change.

Reviewer 3: > well, yes, but that's out of scope for this particular doc ..... I may submit a PR there, if I can find the time -- if anyone else gets to it first -- great!

Don't worry, that is setuptools stuff. I just added the note to remind myself.
